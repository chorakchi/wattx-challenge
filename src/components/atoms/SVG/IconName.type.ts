// These strings are basically names of the files in the assets folder copy pasted here.
export type IconName =
  | "analysis"
  | "arrowDown"
  | "greenCheck"
  | "home"
  | "info"
  | "logout"
  | "proBadge"
  | "setting";
