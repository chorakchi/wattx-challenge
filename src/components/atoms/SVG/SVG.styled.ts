import styled, { css } from "styled-components";
import { ReactSVG } from "react-svg";

export const SVGStyled = styled(ReactSVG)`
  display: inline-flex;
  vertical-align: middle;
  color: ${({ color }) => color || "back"};
  svg,
  span {
    display: flex;
    fill: currentColor;
    ${({ size }) => {
      switch (size) {
        case "small":
          return css`
            width: 18px;
            height: 18px;
          `;
        case "medium":
          return css`
            width: 28px;
            height: 28px;
          `;
        case "large":
          return css`
            width: 42px;
            height: 42px;
          `;
        case "xLarge":
          return css`
            width: 82px;
            height: 82px;
          `;
        default:
          return css`
            width: 100%;
            height: 100%;
          `;
      }
    }}
  }

  path {
    fill: currentColor;
  }
`;
