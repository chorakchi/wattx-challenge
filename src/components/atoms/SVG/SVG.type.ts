import React from 'react';
import { IconName } from './IconName.type';

export type Size = 'xSmall' | 'small' | 'medium' | 'large' | 'xLarge' | 'auto' ;

interface SVGIconProps extends React.SVGAttributes<HTMLDivElement> {

    /**
     * The name of SVG placed in icon folder.
     * don't set the value to name and src at the same time,
     * Just one of them should have value.
     */
    name: IconName;
    /**
     * the address of SVG(art-board)
     * e.g.: "/assets/images/name.svg"
     */
    src?: string;
    /**
     *  "extra small" 10px, <br/>
     *  "small" 16px,<br/>
     *  "base" 24px,<br/>
     *  "medium" 32px,<br/>
     *  "large" 40px,<br/>
     *  "auto": 100%(default)
     * @default 'auto'
     */
    size?: Size;
    iconsFolder?: string;
}

interface SVGSourceProps extends React.SVGAttributes<HTMLDivElement> {
    /**
     * The name of SVG placed in icon folder.
     * don't set the value to name and src at the same time,
     * Just one of them should have value.
     */
    name?: IconName;
    /**
     * the address of SVG(art-board)
     * e.g.: "/assets/images/name.svg"
     */
    src: string;
    /**
     *  "extra small" 10px, <br/>
     *  "small" 16px,<br/>
     *  "base" 24px,<br/>
     *  "medium" 32px,<br/>
     *  "large" 40px,<br/>
     *  "auto": 100%(default)
     * @default 'auto'
     */
    size?: Size;
    iconsFolder?: string;
}

// TS will help the component understand which one to pick based on the properties passed by the parent.
// Eg. if dev passes src but not name, it'll pick the SVGSourceProps as type.
// It basically satisfies TS when there are different use cases for required properties in the type.
export type SVGProps = SVGIconProps | SVGSourceProps;
