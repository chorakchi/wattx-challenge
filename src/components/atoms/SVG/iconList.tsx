// This file is only for Storybook purposes, It has and neither should have anything to do with the actual component
// TODO: Find a new more logical home for this file
export const iconList: Array<string> = [
  "analysis",
  "arrowDown",
  "greenCheck",
  "home",
  "info",
  "logout",
  "proBadge",
  "setting",
];
