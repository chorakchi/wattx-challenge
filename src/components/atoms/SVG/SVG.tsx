import { SVGProps } from "./SVG.type";
import { SVGStyled } from "./SVG.styled";

export const SVG: React.FC<SVGProps> = ({
  name,
  src,
  iconsFolder = "/assets/icons/",
  ...props
}) => {
  const sourcePath = src ? src : `${iconsFolder}${name}.svg`;

  return <SVGStyled {...props} src={sourcePath} wrapper="span" />;
};
