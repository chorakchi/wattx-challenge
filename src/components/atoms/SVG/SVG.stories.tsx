import React from "react";
import { Story, Meta } from "@storybook/react/types-6-0";
import _ from "lodash";

import { SVG } from "./SVG";
import { iconList } from "./iconList";
import { Stack } from "..";

const figmaLink =
  "https://www.figma.com/file/7i7w7VdFONfkl29fgA2PIk/icons?node-id=401%3A950";

export default {
  title: "atoms/SVG",
  component: SVG,
  parameters: {
    docs: {
      description: {
        component:
          `This component make inject external SVG storraged as file into the DOM as markup inline<br/><br/>` +
          `<h3>Icon</h3>` +
          `<iframe width= "100%" height= "400px" src="https://www.figma.com/embed?embed_host=astra&url=https://www.figma.com/file/2LokuyVBjWW20P4CcDJGra/Top-Coins?node-id=10%3A584" ></iframe>
        <br/> <br/> <h5> icon name convention:</h5>` +
          " ```categoryName_specifcName_extraName ``` <br/> <br/>" +
          `Every icon has a name, and is composed by a vector element inside a frame. <br/>
        In order to mantain consistency you should always download and use the frame that wrap the icon from Figma and copy that to icon folder that placed in: <br/>` +
          "``` /src/public/assets/icons ``` <br/> after adding any Icon update the type and list of the icon in ```/src/core/components/atomes/SVG/icon.type.tsx &  iconList.tsx``` (Recommendation: don't copy one by one just select all file copy, paste and modify them in file) <br/><br/>" +
          `for other SVGs(art-boeards) use this folder ` +
          "``` /src/public/assets/images ```",
      },
    },
    backgrounds: {
      default: "light",
    },
  },
} as Meta;

const SVGIcon: Story<any> = (args) => <SVG {...args} />;
const SVGSource: Story<any> = (args) => <SVG {...args} />;

const IconList: Story<any> = ({ list, src, ...args }) => (
  <Stack spacing={10}>
    {list.map((item, i) => (
      <div className="stories-icon-block">
        <div className="stories-icon-item" key={i}>
          <SVG
            {...args}
            iconsFolder={src}
            name={item}
            className="stories-icon-svg"
          />
          <small className="stories-icon-text">{[item]}</small>
        </div>
      </div>
    ))}
  </Stack>
);
export const SVG_ = SVGSource.bind({});
SVG_.args = {
  src: "/assets/images/logo.svg",
};
export const Icon_ = SVGIcon.bind({});
Icon_.args = {
  name: "home",
  size: "lg",
};

export const Icons = IconList.bind({});
Icons.args = {
  size: "base",
  list: iconList,
};
