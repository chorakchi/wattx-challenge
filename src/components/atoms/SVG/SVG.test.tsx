import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { SVG } from "./SVG";

jest.mock("./SVG.styled", () => ({
  SVGStyled: "svg",
}));

describe("SVG", () => {
  it("should render the SVG", () => {
    const subject = shallow(
      <div>
        <SVG name="home" size="large" />
      </div>
    );
    expect(subject).toHaveLength(1);
  });

  it("should render correctly and match to the Snapshot", () => {
    const subject = renderer
      .create(
        <div>
          <SVG name="home" size="large" />
        </div>
      )
      .toJSON();
    expect(subject).toMatchSnapshot();
  });
});
