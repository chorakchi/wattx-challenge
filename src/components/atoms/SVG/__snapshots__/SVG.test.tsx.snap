// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVG should render correctly and match to the Snapshot 1`] = `
<div>
  <svg
    size="large"
    src="/assets/icons/home.svg"
    wrapper="span"
  />
</div>
`;
